walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,163 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  /**
+   * Gets the available time slot for new meeting given existing meeting events.
+   * @param events Existing meeting events, assuming all events end before EOD.
+   * @param request Request to set up a new meeting
+   * @return If one or more time slots exists so that both mandatory and optional attendees can attend,
+   *         returns those time slots. Otherwise, returns the time slots that fit just the mandatory attendees.
+   *         Returns empty if duration is greater than a day.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+    if (!isValidParams(events, duration)) {
+      return Collections.unmodifiableList(new ArrayList<>());
+    }
+
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> attendeesWithOptional = new HashSet<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    // No attendees at all.
+    if (attendeesWithOptional.isEmpty()) {
```

Comment:
If there is no attendees, is it invalid param? should we return empty?
Could this be part of "isValidParams" ?

Reply:
hmm according to given test case, we shall return WHOLE_DAY, so i suppose that is the requirement